investigations/Possible_Instrument_View_Improvements.md

```diff
@@ -0,0 +1,67 @@
+
+## Findings
+
+The following report contains results in terms of percentage inclusive time during the loading of an empty GEM instrument in the instrument view:
+
+- 34% of time is spent doing dynamic casts in several places: 
+	- [CompAssemblyActor::CompAssemblyActor](https://github.com/mantidproject/mantid/blob/7fa758ce336c6ac8ad590c6f3476311d28ed54b2/qt/widgets/instrumentview/src/CompAssemblyActor.cpp#L52)
+	- [ParComponentFactory::create](https://github.com/mantidproject/mantid/blob/546b2eccbaf236fd217ccbd21573f88aa0272103/Framework/Geometry/src/Instrument/ParComponentFactory.cpp#L58)
+	- [ComponentActor::isNonDetector](https://github.com/mantidproject/mantid/blob/7fa758ce336c6ac8ad590c6f3476311d28ed54b2/qt/widgets/instrumentview/src/ComponentActor.cpp#L69)
+	- [Component::Component](https://github.com/mantidproject/mantid/blob/master/Framework/Geometry/src/Instrument/Component.cpp#L25)
+- 49% of time is being spent on `InstrumentActor::getInstrument()` with the following breakdown:
+	- 23.41%  [ExperimentInfo::getInstrument()](https://github.com/mantidproject/mantid/blob/master/Framework/API/src/ExperimentInfo.cpp#L218) where most of the time is spent calling `ParComponentFactory::createInstrument`
+	- 11% `std::use_facet`
+	- 6% `ConfigServiceImpl::getString`
+	- 4% `sp_counted_base::release`
+- Of the 25% inclusive time spent on calls to `ComponentActor::getDetector`, 17% is spent calling `InstrumentActor::getInstrument()`.
+
+## Interpretation 
+### Copying Instrument Geometry
+When creating the instrument view, the entire instrument geometry is copied and stored in "Actors", since the component assembly makes use of virtualisation, several dynamic downcasts must be performed in order to determine types. This can be improved in Instrument 2.0 since the type information is stored in the `ComponentInfo/DetectorInfo` layers without the need for dynamic casting.
+ 
+### Parametrised Instrument Allocation
+In several scenarios within the Instrument View (loading, switching between unwrapped views and picking) a parametrised version of the instrument is created before attempting to access detectors. Every call to `ExperimentInfo::getInstrument()` results in a call to `ExperimentInfo::makeParameterizedInstrument()`, which in turn is triggered by every call to `InstrumentActor::getInstrument()` or `ComponentActor::getComponent()`. For SANS2D, there are 255 calls to the parametrised instrument constructor, for GEM there are ~130K. Similar figures were recorded for switching instrument representations (unwrapped views). This seems to be, in general terms, a design flaw in Instrument 1.0. However, Instrument 2.0 obviates the need to access parameter maps or even the creation of a parametrised instrument. It therefore follows that any access which replaces this object creation with a table lookup could drastically improve performance.
+
+### Caching Layer Lookup vs Detector Object Access
```

Comment:
We should avoid the term "Caching layer". Since instrument 2.0 in it's current form is not caching. It's the actual origin on the `position`, `rotation` etc information. Effectively, it's no longer caching.